commands/operator-sdk/cmd/scorecard/test_definitions.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scorecard
+
+import (
+	"fmt"
+
+	olmAPI "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Struct definitions
+
+type ScorecardVars struct {
+	client        *client.Client
+	crObj         *unstructured.Unstructured
+	csvObj        *olmAPI.ClusterServiceVersion
+	proxyPod      *v1.Pod
+	timeout       int
+	retryInterval int
+}
+
+type Score struct {
+	earnedPoints  int
+	maximumPoints int
+}
+
+type Test struct {
+	testType    string
+	name        string
+	description string
+	// by having an array of scores, we can more easily add support for multiple CR
+	// testing in a future PR
+	scores     []Score
+	cumulative bool
+	run        func(*Test, ScorecardVars) error
```

Comment:
What do you think of making this an interface type. Where an implementation can define the vars it needs (no need for a Vars Struct then) and then make the functions just method recievers on the structs that are aligned with.

Should we consider exposing the fields of the Test?

Reply:
I think its doable using struct composition without too much messiness. We could declare some interfaces and common types to keep track of the tests, and then make a hierarchy of config types that are used to compose them.

Common test interfaces and types might look something like:
```go
type Test interface {
	GetName() string
	GetDescription() string
	Run(context.Context) TestResult
}
type TestResult struct {
	Test          Test
	EarnedPoints  int
        MaximumPoints int
	Suggestions   []string
	Errors        []error
}
```

Then for test setup and config:
```go
type TestInfo struct {
	Name        string
	Description string
}
// Anything that embeds TestInfo only needs to 
// implement Run to implement the Test interface
func (i TestInfo) GetName() string { return i.Name }
func (i TestInfo) GetDescription() string { return i.Description }

type BasicTestConfig struct {
	Client client.Client
	CR     *unstructured.Unstructured
}

type OLMTestConfig struct {
	Client client.Client
	CR     *unstructured.Unstructured
	CSV    *olmapiv1alpha1.ClusterServiceVersion
}
```

Then we could create a type that implements the `Test` interface for each test:

```go
type CheckSpecTest struct {
	TestInfo
	BasicTestConfig
}

func NewCheckSpecTest(conf BasicTestConfig) *CheckSpecTest {
	return &CheckSpecTest{
		BasicTestConfig: conf,
		TestInfo: TestInfo{
			Name:        "Spec Block Exists",
			Description: "Custom Resource has a Spec Block",
		},
	}
}

func (t *CheckSpecTest) Run(ctx context.Context) TestResult {
	res := TestResult{Test: t, MaximumPoints: 1}
	err := t.Client.Get(ctx, types.NamespacedName{Namespace: t.CR.GetNamespace(), Name: t.CR.GetName()}, t.CR)
	if err != nil {
		res.Errors = append(res.Errors, fmt.Errorf("error getting custom resource: %v", err))
		return res
	}
	if t.CR.Object["spec"] != nil {
		res.EarnedPoints++
	}
	if res.EarnedPoints != 1 {
		res.Suggestions = append(res.Suggestions, "Add a 'spec' field to your Custom Resource")
	}
	return res
}
```

Lastly, you could start composing TestSuites and maybe even suites of suites:
```go
type TestSuite struct {
	TestInfo
	Tests       []Test
	TestResults []TestResult

        // Score could be ScoreCumulative or ScorePassFail (or whatever else)
	Score func([]TestResult) (earned, max int)
	
	weights map[string]int
}

func (ts *TestSuite) AddTest(t Test, weight int) {
	ts.Tests = append(ts.Tests, t)
	ts.weights[t.GetName()] = weight
}

func (ts *TestSuite) Run(ctx context.Context) TestResult {
        // 1. Run each of ts.Tests, store each result
        // 2. Aggregate suggestions and errors
        // 3. Calculate a score for the suite using ts.Score and ts.weights
        // 4. Return the aggregated TestResult
}

func NewCumulativeTestSuite(name, description string) *TestSuite {
	return &TestSuite{
		TestInfo: TestInfo{
			Name:        name,
			Description: description,
		},
		Score:   ScoreCumulative,
		weights: make(map[string]int),
	}
}

func NewBasicTestSuite(conf BasicTestConfig) *TestSuite {
	ts := NewCumulativeTestSuite(
		"Basic Tests",
		"Test suite that runs basic, functional operator tests",
	)
	ts.AddTest(NewCheckSpecTest(conf), 44)
	ts.AddTest(NewCheckStatTest(conf), 28)
	ts.AddTest(NewWritingIntoCRsHasEffectTest(conf), 28)

	return ts
}
```

Reply:
The above is more or less what I was imagining as well

Reply:
@joelanford Thanks. I like that design (and I forgot about struct composition, which helps a lot here). I'll change this PR to use a design like the one you described.